File: app/src/main/java/com/dertyp7214/appstore/adapter/AppAdapter.java
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.adapter;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.airbnb.lottie.LottieAnimationView;
import com.dertyp7214.appstore.R;
import com.dertyp7214.appstore.Utils;
import com.dertyp7214.appstore.items.SearchItem;
import com.dertyp7214.appstore.screens.AppScreen;

import java.util.List;

public class AppAdapter extends RecyclerView.Adapter<AppAdapter.ViewHolder> {

    private Activity context;
    private List<SearchItem> appItemList;

    public AppAdapter(Activity context, List<SearchItem> appItemList){
        this.appItemList=appItemList;
        this.context=context;
    }

    @NonNull
    @Override
    public AppAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.app_item_vertical, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull AppAdapter.ViewHolder holder, int position) {
        SearchItem item = appItemList.get(position);

        holder.appTitle.setText(item.getAppTitle());
        holder.appIcon.setImageDrawable(item.getAppIcon());

        for(SearchItem searchItem : appItemList)
            Utils.appsList.put(searchItem.getId(), searchItem);

        holder.view.setOnClickListener(v -> {
            Pair<View, String> icon = Pair.create(holder.appIcon, "icon");
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(context, icon);
            context.startActivity(new Intent(context, AppScreen.class).putExtra("id", item.getId()), options.toBundle());
        });

    }

    class ViewHolder extends RecyclerView.ViewHolder {

        LottieAnimationView appIcon;
        TextView appTitle;
        View view;

        ViewHolder(View itemView) {
            super(itemView);

            appIcon = itemView.findViewById(R.id.appIcon);
            appTitle = itemView.findViewById(R.id.appTitle);
            view = itemView.findViewById(R.id.view);

        }
    }

    @Override
    public int getItemCount() {
        return appItemList.size();
    }
}

File: app/src/main/java/com/dertyp7214/appstore/adapter/AppAdapter.kt
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.adapter

import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.dertyp7214.appstore.R
import com.dertyp7214.appstore.Utils
import com.dertyp7214.appstore.items.SearchItem
import com.dertyp7214.appstore.screens.AppScreen

class AppAdapter(private val context: Activity, private val appItemList: List<SearchItem>) : RecyclerView.Adapter<AppAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppAdapter.ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.app_item_vertical, parent, false))
    }

    override fun onBindViewHolder(holder: AppAdapter.ViewHolder, position: Int) {
        val item = appItemList[position]

        holder.appTitle.text = item.appTitle
        holder.appIcon.setImageDrawable(item.appIcon)

        for (searchItem in appItemList)
            if (!Utils.appsList.containsKey(searchItem.id))
                Utils.appsList[searchItem.id] = searchItem

        holder.view.setOnClickListener {
            val icon = Pair.create<View, String>(holder.appIcon, "icon")
            val options = ActivityOptions.makeSceneTransitionAnimation(context, icon)
            context.startActivity(Intent(context, AppScreen::class.java).putExtra("id", item.id), options.toBundle())
        }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var appIcon: ImageView = itemView.findViewById(R.id.appIcon)
        var appTitle: TextView = itemView.findViewById(R.id.appTitle)
        var view: View = itemView.findViewById(R.id.view)
    }

    override fun getItemCount(): Int {
        return appItemList.size
    }
}
